packages/nextjs/src/server/clerkMiddleware.ts

```diff
@@ -91,82 +94,88 @@ export const clerkMiddleware: ClerkMiddleware = withLogger('clerkMiddleware', lo
-    } else if (requestState.status === AuthStatus.Handshake) {
-      throw new Error('Clerk: handshake status without redirect');
-    }
-
-    const authObject = requestState.toAuth();
-    logger.debug('auth', () => ({ auth: authObject, debug: authObject.debug() }));
-
-    const redirectToSignIn = createMiddlewareRedirectToSignIn(clerkRequest);
-    const protect = createMiddlewareProtect(clerkRequest, authObject, redirectToSignIn);
-    const authObjWithMethods: ClerkMiddlewareAuthObject = Object.assign(authObject, { protect, redirectToSignIn });
-
-    let handlerResult: Response = NextResponse.next();
-    try {
-      handlerResult = (await handler?.(() => authObjWithMethods, request, event)) || handlerResult;
-    } catch (e: any) {
-      handlerResult = handleControlFlowErrors(e, clerkRequest, requestState);
-    }
+    const nextMiddleware: NextMiddleware = async (request, event) => {
+      const clerkRequest = createClerkRequest(request);
+      logger.debug('options', options);
+      logger.debug('url', () => clerkRequest.toJSON());
 
-    // TODO @nikos: we need to make this more generic
-    // and move the logic in clerk/backend
-    if (requestState.headers) {
-      requestState.headers.forEach((value, key) => {
-        handlerResult.headers.append(key, value);
-      });
-    }
+      const requestState = await clerkClient().authenticateRequest(
+        clerkRequest,
+        createAuthenticateRequestOptions(clerkRequest, options),
+      );
 
-    if (isRedirect(handlerResult)) {
-      logger.debug('handlerResult is redirect');
-      return serverRedirectWithAuth(clerkRequest, handlerResult, options);
-    }
+      logger.debug('requestState', () => ({
+        status: requestState.status,
+        headers: JSON.stringify(Object.fromEntries(requestState.headers)),
+        reason: requestState.reason,
+      }));
+
+      const locationHeader = requestState.headers.get(constants.Headers.Location);
+      if (locationHeader) {
+        return new Response(null, { status: 307, headers: requestState.headers });
+      } else if (requestState.status === AuthStatus.Handshake) {
+        throw new Error('Clerk: handshake status without redirect');
+      }
+
+      const authObject = requestState.toAuth();
+      logger.debug('auth', () => ({ auth: authObject, debug: authObject.debug() }));
+
+      const redirectToSignIn = createMiddlewareRedirectToSignIn(clerkRequest);
+      const protect = createMiddlewareProtect(clerkRequest, authObject, redirectToSignIn);
+      const authObjWithMethods: ClerkMiddlewareAuthObject = Object.assign(authObject, { protect, redirectToSignIn });
+
+      let handlerResult: Response = NextResponse.next();
+      try {
+        handlerResult =
+          (await clerkMiddlewareRequestDataStore.run(
+            options,
+            async () => await handler?.(() => authObjWithMethods, request, event),
+          )) || handlerResult;
```

Comment:
This logic is getting a bit hard to follow, maybe we can split it apart for clarity?

```suggestion
        const userHandlerResult = await clerkMiddlewareRequestDataStore.run(
          options,
          async () => handler?.(() => authObjWithMethods, request, event),
        )
        handlerResult = userHandlerResult || handlerResult
```